tls_codec/src/quic_vec.rs

```diff
@@ -0,0 +1,339 @@
+//! # Variable length vectors
+//!
+//! While the TLS RFC 8446 only specifies vectors with fixed length length fields
+//! the QUIC RFC 9000 defines a variable length integer encoding.
+//!
+//! Note that we require, as the MLS specification does, that vectors have to
+//! use the minimum number of bytes necessary for the encoding.
+//! This ensures that encodings are unique.
+
+use alloc::vec::Vec;
+
+use crate::{Deserialize, Error, Serialize, Size};
+
+const MAX_LEN: u64 = 0x3fff_ffff_ffff_ffff; // <= (1<<62)-1
+
+/// Read the length of a variable-length vector.
+///
+/// This function assumes that the reader is at the start of a variable length
+/// vector and returns an error if there's not a single byte to read.
+///
+/// The length and number of bytes read are returned.
+#[inline]
+fn read_variable_length<R: std::io::Read>(bytes: &mut R) -> Result<(usize, usize), Error> {
+    // The length is encoded in the first two bits of the first byte.
+    let mut len_len_byte = [0u8; 1];
+    if bytes.read(&mut len_len_byte)? == 0 {
+        // Return in case there's nothing to read and this is just an
+        // empty vector.
+        return Ok((0, 0));
+    }
+
+    let mut length: usize = (len_len_byte[0] & 0x3F).into();
+    let len_len = (len_len_byte[0] >> 6).into();
+    debug_assert!(len_len <= 3);
+    if len_len > 3 {
+        return Err(Error::InvalidVectorLength);
+    }
+    for _ in 0..len_len {
+        let mut next = [0u8; 1];
+        bytes.read(&mut next)?;
+        length = (length << 8) + usize::from(next[0]);
+    }
+
+    Ok((length, len_len))
+}
+
+#[inline]
+fn length_encoding_bytes(length: u64) -> usize {
+    debug_assert!(length <= MAX_LEN);
+    if length < 0x40 {
+        1
+    } else if length < 0x3fff {
+        2
+    } else if length < 0x3fff_ffff {
+        4
+    } else {
+        8
+    }
+}
+
+// === (De)Serialize for `Vec<T>` and &[T].
+
+impl<T: Serialize + std::fmt::Debug> Serialize for Vec<T> {
+    #[cfg(feature = "std")]
+    #[inline(always)]
+    fn tls_serialize<W: std::io::Write>(&self, writer: &mut W) -> Result<usize, Error> {
+        self.as_slice().tls_serialize(writer)
+    }
+}
+
+impl<T: Size> Size for Vec<T> {
+    #[cfg(feature = "std")]
+    #[inline(always)]
+    fn tls_serialized_len(&self) -> usize {
+        self.as_slice().tls_serialized_len()
+    }
+}
+
+impl<T: Deserialize> Deserialize for Vec<T> {
+    #[cfg(feature = "std")]
+    #[inline(always)]
+    fn tls_deserialize<R: std::io::Read>(bytes: &mut R) -> Result<Self, Error> {
+        let (length, len_len) = read_variable_length(bytes)?;
+        if length == 0 {
+            // An empty vector.
+            return Ok(Vec::new());
+        }
+
+        let mut result = Vec::new();
```

Comment:
Wouldn't it make sense to use something like `with_capacity((length-len_len)/element.tls_serialized_len())` here?

Reply:
We could. It's a trade-off between allocating unnecessary memory and speeding up things. As long as no one measures performance here and identifies this as a bottleneck I'm conservative and don't allocate unnecessary memory.

Reply:
I see. Do I understand correctly, though, that memory would only be unnecessarily allocated if the encoded length is inaccurate?

Reply:
No, any broken element in the vector will cause the function to abort. The scenario I have in mind here is for example a vector that claims to be extremely long where the first element doesn't deserialize correctly. Then we'd allocate a ton of memory here and bail trying to read the first one.